hubgrep/frontend_blueprint/forms/hosting_service/hosting_services.py

```diff
@@ -0,0 +1,164 @@
+import json
+import logging
+
+import markdown
+
+from flask_wtf import FlaskForm
+from wtforms import StringField, SelectField, TextAreaField
+from wtforms.widgets.html5 import URLInput
+from wtforms.validators import ValidationError, URL
+from wtforms.validators import InputRequired
+
+from hubgrep.lib.hosting_service_interfaces import hosting_service_interface_mapping
+from hubgrep.models import HostingService
+
+from hubgrep.frontend_blueprint.forms.hosting_service.validators import (
+    validate_custom_config,
+    validate_url,
+    validate_api_url,
+)
+
+
+logger = logging.getLogger(__name__)
+
+
+class NoHostingServiceFormException(Exception):
+    pass
+
+
+class HostingServiceFormFirstStep(FlaskForm):
+    """ Only get type and landingpage from the user, the rest we automatically resolve to pre-fill in step 2. """
+
+    type = SelectField(
+        u"Type",
+        [InputRequired()],
+        choices=[
+            ("gitea", "gitea"),
+            ("gitlab", "Gitlab"),
+            ("github", "GitHub"),
+        ],
+    )
+    landingpage_url = StringField(
+        "Landingpage Url",
+        [InputRequired(), URL(), validate_url],
+        widget=URLInput(),
+        description="the landingpage for this hoster"
+    )
+
+
+class HostingServiceForm(HostingServiceFormFirstStep):
+    """ Final step when adding a new hosting-service, doubling as the form used for editing hosting-services. """
+
+    # validate_url sanitizes url before validate_api_url makes a call to it
+    api_url = StringField(
+        "Api Url",
+        [InputRequired(), URL(), validate_url, validate_api_url],
+        widget=URLInput(),
+        description="should be the same as the landing page in most cases"
+    )
+    custom_config = TextAreaField(
+        "Custom config",
+        [validate_custom_config],
+        description="Custom config json. If you dont know what to put in, just leave it empty.",
+    )
+    help_text_md = ""
+
+    def populate_api_url(self):
+        self.api_url.data = self.landingpage_url.data
```

Comment:
api urls are the same as landingpage urls - and if not, we have no chance finding out. so, just prefill the landingpage url as api url.

Reply:
landingpage_url is something that should be optionally filled out, isnt it? like we use it for the hosters_list in the frontend to make the links I think, but thats about it?

Reply:
its the first thing we ask for, along with the type. basically the "main url"

Reply:
but we really need 2 different fields, one for landingpage and one for api

Reply:
yea. thats why we still have 2 different fields. :)